# Projeto Técnicas de programação I

## Análise exploratória de dados/Pandas

Nomes: Christyan Costa De Oliveira, Kelvin Pichinini, Luis Ernesto Beingolea Delgado

Turma: 970 LM Tech Data Talents


Nesse projeto trabalharemos com dados fictícios de investidores.

São 3 arquivos que devem ser carregados para responder as perguntas abaixo.

Para a entrega, coloque o notebook respondido no seu github (um por grupo)!

Informações

Cadastro:
- **id** : Identificação do investidor
- **Nome** : Nome completo do investidor 
- **Sexo** : Feminino ou masculino
- **Nasc** : Data de nascimento
- **Idade** : Idade do investidor 
- **Email** : E-mail do investidor
- **Lat_long** : Dados geográficos do investidor
- **Job** : Profisão do investidor ou do seu responsável (em inglês)
- **Estado** : Estado onde o investidor reside

Investimento:
- **id** : Identificação do investidor
- **Salario** : Renda do investidor ou seu responsável
- **Perfil** : Perfil de investimento
- **Ativo01, Ativo02, Ativo03, Ativo04, Ativo05, Ativo06, Ativo07** : O quanto de dinheiro o investidor colocou em cada ativo
- **Nota1, Nota2** : O quanto o investidor está satisfeito com nossa empresa (2 avaliações)

Lembrando que esses dados são irreais e fictícios


Arquivo cadastro.csv


Arquivo investimento.csv (dividido em 2 partes)

### Critérios de avaliação

Os seguintes itens serão avaliados:

1. Reprodutibilidade do código: seu código será executado e precisa gerar os mesmos resultados apresentados por você;

2. Clareza: seu código precisa ser claro e deve existir uma linha de raciocínio direta. Comente o código em pontos que julgar necessário para o entendimento total;

3. Explicação de todos os processos feitos e decisões tomadas

## Informações Gerais sobre o Projeto

- O projeto poderá ser desenvolvido __individualmente__ ou em __grupos com até 4 pessoas__, caos façam em grupos enviar a relação de alunos do grupo para o professor;


- Data de Entrega: 14/04/2023;

- Entrega: Através do _Class_





## 1 - Carregar os arquivos e cadastro, investimento1 e investimento2 e mostrar a quantidade de linhas de cada um, além das informações desses dataframes

In [2]:
import pandas as pd

dfcadastro = pd.read_csv("cadastro.csv", sep=",")
dfinvest1 = pd.read_csv("investimento_parte1.csv", sep=",")
dfinvest2 = pd.read_csv("investimento_parte2.csv", sep=",")

#### Arquivo Cadastro.csv

In [3]:
print("Cadastro:",dfcadastro.shape[0], "linhas;", dfcadastro.shape[1], "colunas\n")
display(dfcadastro.info())

Cadastro: 38754 linhas; 9 colunas

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 38754 entries, 0 to 38753
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   id        38754 non-null  object
 1   Nome      38754 non-null  object
 2   Sexo      38754 non-null  object
 3   Nasc      38754 non-null  object
 4   Idade     38754 non-null  int64 
 5   Email     38754 non-null  object
 6   Lat_long  38754 non-null  object
 7   Job       38754 non-null  object
 8   Estado    38754 non-null  object
dtypes: int64(1), object(8)
memory usage: 2.7+ MB


None

#### Arquivo Investimento_parte1.csv

In [169]:
print("investimento_parte1:",dfinvest1.size, "linhas;", dfinvest1.shape[1], "colunas\n")
display(dfinvest1.info())

investimento_parte1: 240000 linhas; 12 colunas

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 12 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   id       20000 non-null  object 
 1   Salario  20000 non-null  object 
 2   Perfil   20000 non-null  object 
 3   Ativo01  15012 non-null  float64
 4   Ativo02  16857 non-null  float64
 5   Ativo03  16760 non-null  float64
 6   Ativo04  13799 non-null  float64
 7   Ativo05  16829 non-null  float64
 8   Ativo06  9895 non-null   float64
 9   Ativo07  20000 non-null  float64
 10  Nota1    18469 non-null  float64
 11  Nota2    18954 non-null  float64
dtypes: float64(9), object(3)
memory usage: 1.8+ MB


None

#### Arquivo Investimento_parte2.csv

In [170]:
print("investimento_parte1",dfinvest2.size, "linhas;", dfinvest2.shape[1], "colunas\n")
display(dfinvest2.info())

investimento_parte1 288624 linhas; 12 colunas

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24052 entries, 0 to 24051
Data columns (total 12 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   id       24052 non-null  object 
 1   Salario  24052 non-null  object 
 2   Perfil   24052 non-null  object 
 3   Ativo01  17994 non-null  float64
 4   Ativo02  20245 non-null  float64
 5   Ativo03  20249 non-null  float64
 6   Ativo04  16685 non-null  float64
 7   Ativo05  20211 non-null  float64
 8   Ativo06  11932 non-null  float64
 9   Ativo07  24052 non-null  float64
 10  Nota1    22172 non-null  float64
 11  Nota2    22810 non-null  float64
dtypes: float64(9), object(3)
memory usage: 2.2+ MB


None

## 2 - Juntar os dataframes de investimento1 e investimento2 em um só e ver a quantidade de linhas

In [171]:
dfinvest = pd.concat([dfinvest1, dfinvest2])
total_linhas = dfinvest.shape[0]
print(total_linhas,"linhas")


44052 linhas


## 3 - Crie um novo dataframe eliminando as linhas duplicadas

In [172]:
dflimpo = dfinvest.drop_duplicates()
total_sem_duplicadas = dflimpo.shape[0]
print(f'Foram removidas {total_linhas - total_sem_duplicadas} entradas duplicadas')
print('Restando',total_sem_duplicadas,"linhas na base de dados")

Foram removidas 5298 entradas duplicadas
Restando 38754 linhas na base de dados


## 4 - Junte o dataframe de investimento com o de cadastro
Dica: use a chave id

In [173]:
# Inner Join
dfunidos=pd.merge(left = dfcadastro, right = dflimpo, how = 'inner', on = 'id')
print(dfunidos.shape)
display(dfunidos.head(2))

(38754, 20)


id                             Nome  Sexo        Nasc  Idade  \
0  id0000001            Eunice Hellen do Vale   Fem  1970-11-12     51   
1  id0000002  Sebastião Castelo Branco Júnior  Masc  2019-12-10      2   

                    Email                                        Lat_long  \
0        djones@yahoo.com  (Decimal('24.7024355'), Decimal('-28.038258'))   
1  barnesandrea@yahoo.com  (Decimal('21.8589875'), Decimal('122.830566'))   

                    Job Estado Salario    Perfil  Ativo01  Ativo02  Ativo03  \
0      Technical author     TO  $63.67  Moderado  1174.27  4980.98  2638.24   
1  Biochemist, clinical     DF  $21.96  Moderado   842.12  5646.09  3336.07   

   Ativo04  Ativo05  Ativo06  Ativo07      Nota1      Nota2  
0      NaN  4480.91  2329.72   480.72  79.600372   9.802620  
1      NaN  5624.22  2669.98   504.17  44.406893  90.512871

## 5 - Ordene o dataframe pelo id, de forma crescente e redefina o índice

In [174]:
dfunidos.sort_values('id',inplace = True)
dfunidos.set_index('id', inplace=True)

## 6 - Faça uma análise dos dados faltantes e trate-os

Mostre todas as colunas com dados faltantes, sua quantidade/porcentagem e como tratar (preencher) cada uma delas explicando sua estratégia

#### Verificando valores ausentes

In [175]:
data = { 'Total faltante':dfunidos.isnull().sum(),
'Porcentagem faltante':dfunidos.isnull().mean().round(4).mul(100)
}
newdf = pd.DataFrame(data)
display(newdf)

Total faltante  Porcentagem faltante
Nome                   0                  0.00
Sexo                   0                  0.00
Nasc                   0                  0.00
Idade                  0                  0.00
Email                  0                  0.00
Lat_long               0                  0.00
Job                    0                  0.00
Estado                 0                  0.00
Salario                0                  0.00
Perfil                 0                  0.00
Ativo01             9724                 25.09
Ativo02             6100                 15.74
Ativo03             6169                 15.92
Ativo04            11887                 30.67
Ativo05             6191                 15.98
Ativo06            19514                 50.35
Ativo07                0                  0.00
Nota1               3000                  7.74
Nota2               2000                  5.16

A partir da tabela acima é possível verificar que apenas alguns ativos e notas de nps possuem valores ausentes, portanto estes casos serão separados, analisados e receberão tratamentos diferentes.

### Tratamento valores ausentes nas colunas de ATIVOS

Foram levantadas duas hipóteses sobre os motivos pelos quais estes valores estariam ausentes:

H1: Os valores nulos representam na verdade a ausência de valores investidos em um determinado ativo,e, portanto, equivalem ao valor de 0 unidades monetárias.

H2: Falha no processo de preenchimento, armazenamento ou coleta das informações. Neste caso os campos ausentes podem assumir qualquer valor e o mais correto seria excluí-los ou investigar com mais profundidade os motivos e como corrigi-los junto com outras áreas da empresa.

A partir das análises abaixo é possível concluir que nenhuma das colunas de ativos (ativo01-ativo07) possuem registros com o valor = 0. Este é um forte indicativo que de fato os campos nulos representam a ausência de investimento no ativo e portanto pode ser substituído pelo número 0. Está será a hipotese aceita e abordagem adotada neste relatório.


In [176]:
print('Número de registros com o valor= 0 em pelo menos 1 dos ativos: ',
    dfunidos[(dfunidos['Ativo01'] == 0) |
    (dfunidos['Ativo02'] == 0) |
    (dfunidos['Ativo03'] == 0) |
    (dfunidos['Ativo04'] == 0) |
    (dfunidos['Ativo05'] == 0) |
    (dfunidos['Ativo06'] == 0) |
    (dfunidos['Ativo07'] == 0)
].shape[0])

Número de registros com o valor= 0 em pelo menos 1 dos ativos:  0


Preenchendo os campos nulos de ativos com o valor 0

In [177]:
dfunidos.loc[:,'Ativo01':'Ativo07'] = dfunidos.loc[:,'Ativo01':'Ativo07'].fillna(value=0)
display(dfunidos.loc[:,'Ativo01':'Ativo07'].head(2))
data = { 'Total faltante':dfunidos.isnull().sum(),
'Porcentagem faltante':dfunidos.isnull().mean().round(4).mul(100)
}
newdf = pd.DataFrame(data)
display(newdf.loc['Ativo01':'Ativo07'])

Ativo01  Ativo02  Ativo03  Ativo04  Ativo05  Ativo06  Ativo07
id                                                                      
id0000001  1174.27  4980.98  2638.24      0.0  4480.91  2329.72   480.72
id0000002   842.12  5646.09  3336.07      0.0  5624.22  2669.98   504.17

Total faltante  Porcentagem faltante
Ativo01               0                   0.0
Ativo02               0                   0.0
Ativo03               0                   0.0
Ativo04               0                   0.0
Ativo05               0                   0.0
Ativo06               0                   0.0
Ativo07               0                   0.0

### Tratamento valores ausentes nas colunas de NOTAS

In [178]:
display(newdf.loc['Nota1':'Nota2'])
print('Número de registros com ausencia de NOTA1 E NOTA2: ',
    dfunidos[(dfunidos['Nota2'].isna()) & (dfunidos['Nota1'].isna())].shape[0])

Total faltante  Porcentagem faltante
Nota1            3000                  7.74
Nota2            2000                  5.16

Número de registros com ausencia de NOTA1 E NOTA2:  165


Existem 3.000 dados faltantes na coluna NOTA1 e 2.000 faltantes em NOTA2. Estes valores correspondem respectivamente a 7.74% e 5.16% do total de registros. Duas possibilidades para tratar estes dados foram levantadas:

1) Excluir os registros com estes dados faltantes. Com isso seriam excluidos 4365 registros (3000 + 2000 - 165), aproximadamente 13% do total de registros.

2) Criar um modelo para inserir estes dados de modo que as medidas de tendência central e de dispersão permaneçam inalteradas.

In [ ]:
## DECIDIR E REALIZAR TRATAMENTO DAS NOTAS AQUI

## 7 - Manipule os dados a seguir:
- Criar coluna "Total" somando o valor de todos os ativos por investidor
- Criar coluna "Media" com a média das notas 1 e 2
- Criar coluna "Dominio" com apenas o domínio do email (dica: separar por @)
- Mudar o nome da coluna "Salário" para "Renda" e tratar a coluna (sem caracter $ e tipo numérica)
- Mudar os dados da coluna "Sexo" para Masculino e Feminino (tente usar o map)
- Transformar os dados da coluna Media em inteiro
- Criar coluna "NPS" onde o investidor é promotor se a média das notas é maior igual a 80, neutro se a média estiver entre 50 e 80, e detrator se for abaixo de 50


In [162]:
#Criando coluna Total
dfunidos['Total'] = dfunidos[dfunidos.columns[dfunidos.columns.get_loc('Ativo01'):1+dfunidos.columns.get_loc('Ativo07')]].sum(axis=1)
#Criando coluna Média
dfunidos['Media'] = dfunidos[['Nota1','Nota2']].mean(axis=1)
#Criando coluna Dominio
dfunidos['Dominio'] = dfunidos['Email'].str.split('@', expand = True)[1]
#Renomeando coluna salário e removendo sinal '$'
dfunidos.rename(columns={'Salario': 'Renda'}, inplace = True)
dfunidos['Renda'] = dfunidos['Renda'].str.split('$', expand = True)[1].astype(float)
#Mudar os dados da coluna "Sexo" para Masculino e Feminino (tente usar o map)
dicio = {"Masc": "Masculino", "Fem": "Feminino"}
dfunidos['Sexo'] = dfunidos['Sexo'].map(dicio)
#Criar coluna "NPS" onde o investidor é promotor se a média das notas é maior igual a 80, neutro se a média estiver entre 50 e 80, e detrator se for abaixo de 50
def nps(media):
  if media >=80:
    return 'Promotor'
  elif media>=50:
    return 'Neutro'
  else:
    return 'Detrator'
dfunidos['NPS'] = dfunidos['Media'].map(nps)
dfunidos.info()
display(dfunidos.head(2))

<class 'pandas.core.frame.DataFrame'>
Index: 38754 entries, id0000001 to id0038754
Data columns (total 23 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Nome      38754 non-null  object 
 1   Sexo      38754 non-null  object 
 2   Nasc      38754 non-null  object 
 3   Idade     38754 non-null  int64  
 4   Email     38754 non-null  object 
 5   Lat_long  38754 non-null  object 
 6   Job       38754 non-null  object 
 7   Estado    38754 non-null  object 
 8   Renda     38754 non-null  float64
 9   Perfil    38754 non-null  object 
 10  Ativo01   38754 non-null  float64
 11  Ativo02   38754 non-null  float64
 12  Ativo03   38754 non-null  float64
 13  Ativo04   38754 non-null  float64
 14  Ativo05   38754 non-null  float64
 15  Ativo06   38754 non-null  float64
 16  Ativo07   38754 non-null  float64
 17  Nota1     35754 non-null  float64
 18  Nota2     36754 non-null  float64
 19  Total     38754 non-null  float64
 20  Media     38589 non-n

Nome       Sexo        Nasc  Idade  \
id                                                                         
id0000001            Eunice Hellen do Vale   Feminino  1970-11-12     51   
id0000002  Sebastião Castelo Branco Júnior  Masculino  2019-12-10      2   

                            Email  \
id                                  
id0000001        djones@yahoo.com   
id0000002  barnesandrea@yahoo.com   

                                                 Lat_long  \
id                                                          
id0000001  (Decimal('24.7024355'), Decimal('-28.038258'))   
id0000002  (Decimal('21.8589875'), Decimal('122.830566'))   

                            Job Estado  Renda    Perfil  ...  Ativo04  \
id                                                       ...            
id0000001      Technical author     TO  63.67  Moderado  ...      0.0   
id0000002  Biochemist, clinical     DF  21.96  Moderado  ...      0.0   

           Ativo05  Ativo06  Ativo07      Nota1      Nota2     Total  \
id                                                                     
id0000001  4480.91  2329.72   480.72  79.600372   9.802620  16084.84   
id0000002  5624.22  2669.98   504.17  44.406893  90.512871  18622.65   

               Media    Dominio       NPS  
id                                         
id0000001  44.701496  yahoo.com  Detrator  
id0000002  67.459882  yahoo.com    Neutro  

[2 rows x 23 columns]

In [163]:
#Criando coluna de média das notas
#dfunidos['Media'] = dfunidos['Media'].astype(int)

#PRECISA IMPLEMENTAR TRATAMENTO DAS NOTAS ANTES!!!


display(dfunidos)


Nome       Sexo        Nasc  Idade  \
id                                                                            
id0000001               Eunice Hellen do Vale   Feminino  1970-11-12     51   
id0000002     Sebastião Castelo Branco Júnior  Masculino  2019-12-10      2   
id0000003           Romildo Fábio de Carvalho  Masculino  2015-12-28      6   
id0000004           Lana Albina de Lemos Mumy   Feminino  1993-05-03     28   
id0000005    Rose Elisângela Vyrnium de Moura   Feminino  1984-06-10     37   
...                                       ...        ...         ...    ...   
id0038750     Beatriz Maísa de Miranda Milani   Feminino  1999-02-20     22   
id0038751                     Juan dos Santos  Masculino  1999-07-10     22   
id0038752       Severino Edson de Viana Loase  Masculino  1981-12-01     40   
id0038753                          Kim Mooshi  Masculino  1999-12-06     22   
id0038754  Vinícius Arnaldo de Pereira Mendes  Masculino  2003-09-13     18   

                             Email  \
id                                   
id0000001         djones@yahoo.com   
id0000002   barnesandrea@yahoo.com   
id0000003  walkermichael@yahoo.com   
id0000004   cobbangela@hotmail.com   
id0000005  crawfordcraig@yahoo.com   
...                            ...   
id0038750         mark47@gmail.com   
id0038751  greggnguyen@hotmail.com   
id0038752        cindy11@yahoo.com   
id0038753  mcdonaldmario@gmail.com   
id0038754    omartinez@hotmail.com   

                                                  Lat_long  \
id                                                           
id0000001   (Decimal('24.7024355'), Decimal('-28.038258'))   
id0000002   (Decimal('21.8589875'), Decimal('122.830566'))   
id0000003   (Decimal('-28.2161525'), Decimal('97.140529'))   
id0000004   (Decimal('-40.8557635'), Decimal('-2.936099'))   
id0000005  (Decimal('-20.2620885'), Decimal('104.349223'))   
...                                                    ...   
id0038750  (Decimal('-23.8494875'), Decimal('-91.561304'))   
id0038751   (Decimal('-7.8876355'), Decimal('155.835880'))   
id0038752  (Decimal('-52.1338005'), Decimal('-59.715325'))   
id0038753  (Decimal('-81.2874995'), Decimal('-28.805800'))   
id0038754     (Decimal('-37.248521'), Decimal('6.033254'))   

                                             Job Estado     Renda  \
id                                                                  
id0000001                       Technical author     TO     63.67   
id0000002                   Biochemist, clinical     DF     21.96   
id0000003  Development worker, international aid     DF  62497.64   
id0000004              Estate manager/land agent     DF  99878.74   
id0000005    Chartered public finance accountant     MS    188.01   
...                                          ...    ...       ...   
id0038750             Further education lecturer     RS   4993.49   
id0038751                  Adult guidance worker     PI  41218.63   
id0038752               Arts development officer     PR     32.41   
id0038753                     Production manager     RJ    914.52   
id0038754                 Pension scheme manager     AL  29076.19   

                Perfil  ...   Ativo04  Ativo05  Ativo06  Ativo07      Nota1  \
id                      ...                                                   
id0000001     Moderado  ...      0.00  4480.91  2329.72   480.72  79.600372   
id0000002     Moderado  ...      0.00  5624.22  2669.98   504.17  44.406893   
id0000003    Agressivo  ...      0.00  5033.83  2229.45   523.12  44.020024   
id0000004    Agressivo  ...      0.00  5933.91     0.00   486.05  25.597924   
id0000005  Conservador  ...  30841.35  6044.86  2198.89   512.98  14.463307   
...                ...  ...       ...      ...      ...      ...        ...   
id0038750     Moderado  ...   8454.99     0.00  2191.78   489.76  66.644257   
id0038751     Moderado  ...  11243.08  4692.11  2368.93   496.02  96.987294   
id0038752    Agressivo 

## 8 - Calcule os seguintes dados
- Calcule o total investido em cada coluna Ativo (tente usar o apply)
- Qual o menor valor investido em cada ativo?
- Mostre a média e desvio padrão de cada ativo
- Qual a idade do investidor mais velho?
- Qual a mediana da coluna Média?

## 9 - Responder as perguntas abaixo

- Quantos investidores são promotores?
- Quantos investidores são do sexo feminino?
- Quantos investidores investiram mais de 16000 no Ativo04?
- Quantos investidores investiram mais de 16000 no Ativo04 e menos de 5000 no Ativo05?
- Quantos investidores do sexo feminino tem mais de 25 anos e investiu menos de 10000? (bonus: por que esse dado seria importante para a estratégia da nossa empresa?

- Quantos investidores tem em cada estado (dica: use groupby)
- Qual é a soma do total investido por cada estado?
- Como os investidores se distribuem em promotores, detratores e neutros?
- Qual é a média do total investido por promotores, detratores e neutros?
